### DISCUSSION TITLE
Fixing random effects in data simulation

### DISCUSSION THREAD

**Discussion by AngeliaMiller** [sdmTMB/sdmTMB#251] [Q&A] [answered]
Hello, 

I am trying to simulate new data using sdmtmb_simulate() and parameters estimated by a sdmTMB model fit. In doing this I would like to fix my spatial random effects estimated from my previously fit model, and simulate new spatiotemporal fields. 

In practice, I know I should use the argument in the function `fixed_re = list(omega_s = [my spatial fields vector], epsilon_st = NULL, zeta_s = NULL)`. However, when I do this, sdmTMB has trouble reading my omega_s vector which has the same number of observations as the data that I am supplying. 

Specifically the error is: 
`Error in EvalDoubleFunObject(Fun, unlist(parameters), get_reportdims = TRUE) : 
  Error when reading the variable: 'omega_s'. Please check data and parameters.`

The only way I am able to supply my spatial fields to `fixed_re` is as `fixed_re = list(omega_s)` but I believe it is still simulating new spatial fields because my output omega_s values are not the same as my input omega_s values. 

Any advice? 

Thanks!

**Comment by seananderson** [comment, marked as answer]
The `omega_s` passed to `sdmTMB_simulate()` need to be at the locations of the mesh vertices, not the data locations. If you're estimating from a fitted model, this could theoretically also be done from `simulate()`, but I just realized I never exposed an argument to pass in which random effects should be simulated as new (I'll try to fix that soon). Here's an example with `sdmTMB_simulate()`:

``` r
library(ggplot2)
theme_set(theme_light())
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
library(sdmTMB)

mesh <- make_mesh(pcod_2011, c("X", "Y"), cutoff = 10)
fit <- sdmTMB(
  density ~ depth_scaled,
  family = tweedie(),
  data = pcod_2011, mesh = mesh,
  time = "year"
)

b <- tidy(fit, "ran_pars")
pars <- get_pars(fit)
# omega_s at mesh vertices, not data locations:
head(pars$omega_s)
#>            [,1]
#> [1,]  1.4817846
#> [2,] -3.5742845
#> [3,] -1.1594887
#> [4,] -0.8141616
#> [5,] -1.5041945
#> [6,]  0.3484860
pred <- predict(fit)

set.seed(12810)
s <- sdmTMB_simulate(
  density ~ depth_scaled,
  mesh = mesh,
  data = pcod_2011,
  time = "year",
  family = tweedie(),
  B = coef(fit), # assumes a recent version of sdmTMB
  phi = exp(pars$ln_phi),
  sigma_E = b$estimate[b$term == "sigma_E"],
  sigma_O = b$estimate[b$term == "sigma_O"],
  range = b$estimate[b$term == "range"],
  fixed_re = list(omega_s = pars$omega_s)
)
head(s)
#>   year        X        Y    omega_s epsilon_st           mu        eta observed
#> 1 2011 435.2374 5717.604  1.4817846   2.286642 329.53103005  5.7976705 59.73576
#> 2 2011 486.7163 5718.942 -3.5742845  -1.247754   0.43596095 -0.8302026  0.00000
#> 3 2011 489.5047 5716.534 -3.1500656  -1.166730   0.82235691 -0.1955808  0.00000
#> 4 2011 544.9725 5717.464 -1.1594887  -1.012765   1.49995699  0.4054364  0.00000
#> 5 2011 403.5942 5720.350 -0.8141616   1.405214   7.23040134  1.9782945  0.00000
#> 6 2011 419.6635 5720.900 -1.5041945  -1.961440   0.09690404 -2.3340340  0.00000
#>   (Intercept) depth_scaled
#> 1           1    0.7408840
#> 2           1   -2.7062455
#> 3           1   -2.9334899
#> 4           1   -0.2224156
#> 5           1    1.8685071
#> 6           1    2.3175211

# same:
ggplot(s, aes(X, Y, colour = omega_s)) +
  geom_point() +
  scale_colour_gradient2()
```

![](https://i.imgur.com/Qovevrw.png)

``` r
ggplot(pred, aes(X, Y, colour = omega_s)) +
  geom_point() +
  scale_colour_gradient2()
```

![](https://i.imgur.com/wJzjKIi.png)

``` r

# different:
filter(s, year %in% 2011) |>
  ggplot(aes(X, Y, colour = epsilon_st)) +
  geom_point() +
  scale_colour_gradient2()
```

![](https://i.imgur.com/kqhLEQG.png)

``` r
filter(pred, year %in% 2011) |>
  ggplot(aes(X, Y, colour = epsilon_st)) +
  geom_point() +
  scale_colour_gradient2()
```

![](https://i.imgur.com/6xLj2Ig.png)

<sup>Created on 2023-08-25 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by AngeliaMiller** [comment]
Thanks for this @seananderson, super helpful! 

What about if I'm simulating data for forecasted years that were included in the model fit? In this case, I have resampled my existing data to serve as the locations for my future years.  By simulating at these "new locations" which are technically still included in the mesh, the code you included gets me a slightly different omega_s plot for my simulations in future years compared to my predictions for the resampled years. Some of my code for this is below. 

Is it also possible to simulate data for new locations that are not included in the mesh, say for instance at our grid centroid points, while fixing with the predicted spatial fields? 

```
# read in previous fit model
fit_preds <- predict(fall_mod)
ran_pars <- tidy(fall_mod, "ran_pars")
all_pars <- get_pars(fall_mod)

# randomly select 5 years 
resampled_years <- sample(c(2009:2019,2021), size = 5, replace = TRUE)

# use locations from the resampled years as the observations in future years
# loop to filter each resampled year individually and bind together into empty storage filter for instances where a year is replaced and resampled
future_preds <- data.frame() # empty storage dataframe for filtered tow locations

for(t in seq_along(resampled_years)){
  filtered_yr <- fit_preds |> 
    filter(EST_YEAR %in% resampled_years[t]) |> 
    mutate(EST_YEAR = future_years[t])
  
  future_preds <- bind_rows(future_preds, filtered_yr) 

}

# make mesh of "new" locations, which could still be found in the mesh used to fit the original model 
mesh <- make_mesh(future_preds, xy_cols = c("X", "Y"), cutoff = 10)

base.sim <- sdmTMB_simulate(
  EXPCATCHWT ~ poly(AVGDEPTH, 2) + as.factor(AREA),
  data = data,
  mesh = mesh,
  family = tweedie(link = "log"),
  time = "EST_YEAR",
  B = c(-0.66, -41.8, -25.4, 0.09), # coefficient estimates 
  range = ran_pars$estimate[ran_pars$term == "range"], 
  sigma_O = ran_pars$estimate[ran_pars$term == "sigma_O"],
  sigma_E = ran_pars$estimate[ran_pars$term == "sigma_E"],
  phi = exp(all_pars$ln_phi), # dispersion parameter
  tweedie_p = ran_pars$estimate[ran_pars$term == "tweedie_p"], # tweedie power parameter 
  fixed_re = list(omega_s = all_pars$omega_s, epsilon_st = NULL, zeta_s = NULL))

```

From simulated data for forecasted years: 
```
ggplot(sim_base_tows, aes(X, Y, colour = omega_s)) +
  geom_point() +
  scale_colour_gradient2()
```

<img src ="https://github.com/pbs-assess/sdmTMB/assets/83592135/a1c270cf-ce17-4c43-9271-faf141a7b472" width = "600" />

From predictions for model fit data: 
```
ggplot(fit_preds |> filter(EST_YEAR %in% resampled_years), aes(X, Y, colour = omega_s)) +
  geom_point() +
  scale_colour_gradient2()
```

<img src="https://github.com/pbs-assess/sdmTMB/assets/83592135/298e9323-93ce-479f-9fbd-8f4ca3fc3da4" width = "600" />
